php/libraries/File_Upload_Validator.class.inc

```diff
@@ -0,0 +1,177 @@
+<?php declare(strict_types=1);
+
+/**
+ * Class to upload files in a safe way.
+ *
+ * This class provides validation for files being uploaded to LORIS. It should
+ * serve as the ONLY way that files are uploaded so that all file uploading
+ * functionality can occur on a well-tested foundation.
+ *
+ * PHP Version 7
+ *
+ *  @category Main
+ *  @package  Behavioural
+ *  @author   John Saigle <john.saigle@mcin.ca>
+ *  @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ *  @link     https://www.github.com/aces/Loris
+ */
+
+/**
+ * Class to validate file uploads to LORIS.
+ *
+ *  @category Main
+ *  @package  Behavioural
+ *  @author   John Saigle <john.saigle@mcin.ca>
+ *  @license  http://www.gnu.org/licenses/gpl-3.0.txt GPLv3
+ *  @link     https://www.github.com/aces/Loris
+ */
+class File_Upload_Validator
+{
+
+    public $uploadDirectory;
+    public $permittedMIMETypes;
+    private $_responseCode = 500;
+
+    /**
+     * Create new instance of a File Upload Validator.
+     *
+     * @param string $uploadDirectory    The target parent directory of files
+     *                              uploaded using this instance.
+     * @param array  $permittedMIMETypes A whitelist of all MIME types that can
+     * be accepted by this instance. E.g. array('image/png', 'text/plain')
+     *
+     * @return void
+     */
+    public function __construct(
+        string $uploadDirectory,
+        array $permittedMIMETypes = array()
+    ) {
+        $this->uploadDirectory    = $uploadDirectory;
+        $this->permittedMIMETypes = $permittedMIMETypes;
+        $this->validate();
+    }
+
+    /**
+     * Validates the constructor by ensuring the upload path exists and is
+     * writable.
+     *
+     * @throws \LorisException
+     *
+     * @return void
+     */
+    protected function validate(): void
+    {
+        if (! is_dir($this->uploadDirectory)) {
+            $this->_responseCode = 500;
+            throw new \LorisException(
+                'Upload directory is not a directory!'
```

Comment:
I don't think adding this restriction is a valid assumption, since often uploaders categorize things in some way (such as `dir/candid/filename` or `dir/uploaduser/filename`) and the directory may not exist.